### ISSUE TITLE
Training time

### ISSUE THREAD

**Issue by meltingCat** [BindsNET/bindsnet#431]
MNIST examples readme said that "Training can be as fast as 1 minute/epoch on GPU (1080ti)" , but I trained about 2hrs on RTX2080. Running the command below.
`python batch_eth_mnist.py --gpu --batch_size 32 `

**Comment by SimonInParis** [comment]
Hi. This seems odd.
Does the python script example starts by mentioning:
Running on Device = 'cuda'

Simon CABY

**Comment by meltingCat** [comment]
Yes, it shows "Running on Device = 'cuda'", but GPU usage is 0%
```
    0  GeForce RTX 2080    Off  | 00000000:01:00.0  On |                  N/A |
| 25%   42C    P2    35W / 215W |   1340MiB /  7979MiB |      0%      Default |
```

**Comment by SimonInParis** [comment]
You are right!
Somehow, some recent changes in the plotting system takes the computation
out of the GPU.
Just edit the source and remove plotting (--plot option), and it will run
again on the GPU.

for ex.:
[image: image.png]

Once fixed, and with batch_size=320, I get about 30 seconds/epoch.

Thanks for finding this bug.
We will investigate.

Simon

**Comment by meltingCat** [comment]
@SimonInParis 
Thanks for your reply. I remove plotting (--plot option), and speed up the training, but I get a very low accuracy below with command `python batch_eth_mnist.py --batch_size 320 --n_neurons 100`

```
Running on Device =  cuda

Begin training.

Progress: 0 / 1 (0.0000 seconds)
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 188/188 [00:57<00:00,  3.26it/s]
Progress: 1 / 1 (57.7351 seconds)
Training complete.


Begin testing

Test progress: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉| 9999/10000 [15:14<00:00, 11.16it/s]
All activity accuracy: 0.10
Proportion weighting accuracy: 0.10 

Progress: 1 / 1 (914.7606 seconds)
Testing complete.

Test progress: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [15:14<00:00, 10.93it/s]
```

**Comment by meltingCat** [comment]
I tried to optimum results using 1600 like what you suggest before, but facing the issue below with command `python batch_eth_mnist.py --batch_size 32 --n_neurons 1600`
```
Running on Device =  cuda

Begin training.

Progress: 0 / 1 (0.0000 seconds)
 14%|███████████████████                                                                                                                         | 256/1875 [01:11<07:26,  3.63it/s]
All activity accuracy: 10.21 (last), 10.21 (average), 10.21 (best)
Proportion weighting accuracy: 10.21 (last), 10.21 (average), 10.21 (best)

 14%|███████████████████                                                                                                                         | 256/1875 [01:11<07:33,  3.57it/s]
Progress: 1 / 1 (71.6520 seconds)
Training complete.


Begin testing

Test progress: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉| 9999/10000 [16:08<00:00, 10.19it/s]
All activity accuracy: 0.51
Proportion weighting accuracy: 0.48 

Progress: 1 / 1 (968.5211 seconds)
Testing complete.

Test progress: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [16:08<00:00, 10.32it/s]
```

**Comment by SimonInParis** [comment]
Hi.

I proposed a batch of 320 just to have the fastest running model on GPU,
but very large batches tend to train the model slower, paradoxically.
So there's a balance here.
Usually, you can get very good results with a batch size around 32 (default
value). Lower batch sizes take longer to train but reach slightly higher
results.
After 2 or 3 epochs, you should get over 90% or more.

Batches in SNN are tricky. Learning of each batch is summed, then applied
to the weights. If the learning rate is too high or if the batch size is
too big, it may lead to instability in training.

Simon

**Comment by Hananel-Hazan** [comment]
Thanks @SimonInParis

**Comment by yizexiang** [comment]
> You are right! Somehow, some recent changes in the plotting system takes the computation out of the GPU. Just edit the source and remove plotting (--plot option), and it will run again on the GPU. for ex.: [image: image.png] Once fixed, and with batch_size=320, I get about 30 seconds/epoch. Thanks for finding this bug. We will investigate. Simon
> […](#)

Hi. I run batch_eth_mnist.py using Colab with Tesla K80. Plotting is turned off, but the training speed is very slow, about 10 minute/epoch. I don't where the problem is. Thanks!

**Comment by yizexiang** [comment]
> You are right! Somehow, some recent changes in the plotting system takes the computation out of the GPU. Just edit the source and remove plotting (--plot option), and it will run again on the GPU. for ex.: [image: image.png] Once fixed, and with batch_size=320, I get about 30 seconds/epoch. Thanks for finding this bug. We will investigate. Simon
> […](#)

Hi, @Hananel-Hazan , @SimonInParis. I run batch_eth_mnist.py using Colab with Tesla K80. Plotting is turned off, but the training speed is very slow, about 10 minute/epoch. I don't where the problem is. Thanks!

**Comment by Hananel-Hazan** [comment]
Try to run with different batch sizes, you may find the right size of batch that fit your card specification.

**Comment by yizexiang** [comment]
i'll try it out, thanks
